crates/solvers/src/boundary/naive.rs

```diff
@@ -0,0 +1,239 @@
+                    input,
+                    output,
+                    internalize: false,
+                })
+            })
+            .collect(),
+    })
+}
+
+// Beyond this point is... well... nameless and boundless chaos. The
+// unfathomable horrors that follow are not for the faint of heart!
+//
+// Joking asside, the existing naive solver implementation is tightly coupled
+// with the `Settlement` and `SettlementEncoder` types in the `solver` crate.
+// This meeans that there is no convinient way to say: "please compute a
+// solution given this list of orders and constant product pool" without it
+// creating a full settlement for encoding. In order to adapt that API into
+// something that is useful in this boundary module, we create a fake slippage
+// context that applies 0 slippage (so that we can recover the exact executed
+// amounts from the constant product pool) and we create capturing settlement
+// handler implementations that record the swap that gets added to each
+// settlement so that it can be recovered later to build a solution.
+
+struct Slippage {
+    calculator: SlippageCalculator,
+    prices: ExternalPrices,
+}
+
+impl Slippage {
+    fn new(tokens: liquidity::TokenPair) -> Self {
+        // We don't actually want to include slippage yet. This is because the
+        // Naive solver encodes liquidity interactions and the driver is
+        // responsible for applying slippage to those. Create a dummy slippage
+        // context for use with the legacy Naive solver.
+        let (token0, token1) = tokens.get();
+        Self {
+            calculator: SlippageCalculator::from_bps(0, None),
+            prices: ExternalPrices::new(
+                H160::default(),
+                [
+                    (token0.0, BigRational::one()),
+                    (token1.0, BigRational::one()),
+                ]
+                .into_iter()
+                .collect(),
+            )
+            .unwrap(),
+        }
+    }
+
+    fn context(&self) -> SlippageContext {
+        self.calculator.context(&self.prices)
+    }
+}
+
+struct OrderHandler {
+    order: Order,
+}
+
+impl SettlementHandling<LimitOrder> for OrderHandler {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn encode(&self, execution: U256, encoder: &mut SettlementEncoder) -> anyhow::Result<()> {
+        encoder.add_trade(
+            self.order.clone(),
+            execution,
+            self.order.metadata.solver_fee,
+        )?;
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+struct PoolHandler {
+    swap: Mutex<Option<(eth::Asset, eth::Asset)>>,
+}
+
+impl SettlementHandling<ConstantProductOrder> for PoolHandler {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn encode(
+        &self,
+        execution: AmmOrderExecution,
+        _: &mut SettlementEncoder,
+    ) -> anyhow::Result<()> {
+        let mut swap = self.swap.lock().unwrap();
```

Comment:
nit: You could prepare the value before locking and then simply assign. That would mean nobody could cause the lock to be held for too long by introducing new code.
Again it wouldn't make a difference here. I just want to promote safer access patterns with locks.

Reply:
6700e0fb429776121c5399e4388c24e763c29367